impeller/docs/babys_first_triangle.md

```diff
@@ -0,0 +1,276 @@
+auto desc = PipelineBuilder<VS, FS>::MakeDefaultPipelineDescriptor(*context);
+```
+
+`MakeDefaultPipelineDescriptor` fills in the default values for the fixed function stages for the pipeline. Depending on the environment, some minor tweaks to this descriptor might be necessary.
+
+Astute readers will notice that this is the first time we have tried to fuse the vertex and fragment stages together. But what happens if they aren't compatible? For instance, what if the fragment stage expects an input the vertex stage doesn't provide? The compiler has thought of this for you and the act of putting together a pipeline builder via `PipelineBuilder<VS, FS>` will statically (at build time) perform the check for you. If the stages aren't compatible the C++ code will refuse to compile.
+
+Finally, we have everything we need, let's create the pipeline:
+
+```c++
+auto pipeline = context->GetPipelineLibrary()->GetPipeline(desc).Get();
+```
+
+This call creates a pipeline on a background thread and returns a future. But we need the pipeline right away. Just `Get` wait on the pipeline creation. Remember, pipelines are expensive to create. So just do this once upfront and keep them around for as long as possible. And don't create these in frame workloads unless you want jank.
+
+### Vertex Data
+
+Remember the `PerVertexData` struct the compiler generated for us? It is our job to fill in the the vertex information and provide it as a GPU buffer allocation to the draw call. Handily, there is another C++ utility called the `VertexBufferBuilder<T>` that can create this buffer for us. The template parameter T is the `PerVertexData` struct.
+
+Let's create the data we will be giving our draw call. Remember, our vertex shader is a bit of slacker and doesn't do anything to the vertex information to convert it to normalized device coordinates. So we need to make sure the information is already in [normalized device coordinate](coordinate_system.md).
+
+```c++
+VertexBufferBuilder<VS::PerVertexData> vertex_buffer_builder;
+vertex_buffer_builder.AddVertices({
+    {{-0.5, -0.5}},
+    {{0.0, 0.5}},
+    {{0.5, -0.5}},
+});
+```
+
+Ask the vertex buffer builder to create the device buffer for us. Since we aren't going to be changing the coordinates of the vertex per frame, we can do this once upfront and keep referencing the same buffer over and over in our draw calls.
+
+```c++
+auto vertex_buffer = vertex_buffer_builder.CreateVertexBuffer(
+      *context->GetResourceAllocator());
+```
+
+If you change the vertex information in your shader, the `PerVertexData` struct will change and cause a compile time error where you setup your vertex buffer. This way, you can be confident that refactoring your shaders will immediately flag the instances where you are specifying vertex data to your shader.
+
+### Draw
+
+You don't all the heavy lifting already. Per frame, you only need to set the pipeline and vertex buffer you've stashed in the render pass and invoke a draw call.
```

Comment:
missing verb in first sentence

Reply:
or maybe "You don't" is supposed to be "You've done"?

Reply:
The latter. Fixing.